RFS/src/main/java/org/opensearch/migrations/bulkload/common/OpenSearchClient.java

```diff
@@ -99,17 +129,76 @@ private Version versionFromResponse(HttpResponse resp) {
 
             var distroNode = versionNode.get("distribution");
             if (distroNode != null && distroNode.asText().equalsIgnoreCase("opensearch")) {
-                versionBuilder.flavor(Flavor.OPENSEARCH);
+                versionBuilder.flavor(getLikelyOpenSearchFlavor());
             } else {
                 versionBuilder.flavor(Flavor.ELASTICSEARCH);
             }
-            return versionBuilder.build();
+            return Mono.just(versionBuilder.build());
         } catch (Exception e) {
             log.error("Unable to parse version from response", e);
-            throw new OperationFailed("Unable to parse version from response: " + e.getMessage(), resp);
+            return Mono.error(new OperationFailed("Unable to parse version from response: " + e.getMessage(), resp));
         }
     }
 
+    Mono<Boolean> checkCompatibilityModeFromResponse(HttpResponse resp) {
```

Comment:
Non-blocking nit-pick: May I suggest re-naming this to `isClusterInCompatibilityMode()`?  I think naming methods that answer a boolean question as a question helps make their intent clearer.  The current name is unclear on what the outcome/side-effect of the check is (IMO).